ocl: implement alpha blending

dspmeng opened the pull request:
- import drm prime fd from VASurfaceID to create cl image
- add opencl kernel for alpha blending
- add profiling macro to blend example

There is hard code in vpp/Makefile.am for Beignet libcl.so include path, because Beignet doesn't provide pkg-config .pc file. Any suggestion?

xuguangxin committed:
ocl: init version for opencl abstract

libva's alpha blending have limited support,
it only supports src&dest have same size.
It's inconvenience for real word application.
We'd better enable opencl based vpp functions.
It also a step stone for other vpp operation.
such as rotation,flips...etc

xuguangxin committed:
vpp: add  opencl based alpha blending

it's just a bare-bone application to test all base class works.
it waiting for opencl kernels.

xuguangxin committed:
factory: make sure factory return IVideoXXX interface.

OclPostProcessBlender do not inhert from VaapiPostProcessBase
more over, not all sub class deriver from VaapiDecoderBase or VaapiEncoderBase
so make sure factory return IVideoXXX is better

xuguangxin committed:
example: add alpha blend app

an alpha blending application to render random color blocks on a video.

xuguangxin committed:
interface: change VideoRect.x and VideoRect.y to uint32_t

xuguangxin committed:
tests: prefetch first frame in vppinputdecode

it make sure we can get width height after vppinputdecode configed

xuguangxin committed:
vpp: remove backgroud clean in scaler

we do not need clean back ground for scaler, and it will make whole screen
black when we use NV12 as dest, RGBX have no this problem, because there
are some issue in opensource libva driver.

xuguangxin committed:
baseenc: refact, add createNewSurface to prepare for next commit

xuguangxin committed:
common & vaapi: add surface allocator

xuguangxin committed:
refact: promote videopool to common

xuguangxin committed:
baseenc: pass all reconstruct surface to vaCreateContext

android's libva driver need this

xuguangxin committed:
h264enc: clean reflist0/1 for I frame

we did not clear reflist0/1 for I frame, android driver will crash on this

xuguangxin committed:
interface: decoder, support external allocator

xuguangxin committed:
h264enc: bug fix, init m_frameIndex before use it

else valgrind will complain use uninit value at L922

xuguangxin commented on vpp/Makefile.am:
```diff
@@ -57,7 +57,11 @@ libyami_vpp_cppflags = \
         $(NULL)
 
 if BUILD_OCL_BLENDER
-        libyami_vpp_ldflags += -lOpenCL
```
is any possible to find a portable way, since it maybe not a x64 platform

xuguangxin commented on vpp/oclpostprocess_blender.cpp:
```diff
@@ -25,18 +25,182 @@
+    VAImage src_image, dst_image;
+    VABufferInfo src_info, dst_info;
+
+    if (VA_STATUS_SUCCESS != vaDeriveImage(m_display, src_id, &src_image)) {
```
could we use one function like 
 cl_mem convertToClImage(const SharedPtr<VideoFrame>& frame, cl_image_format format, int plane)  to hide all vaDeriveImage, vaAcquireBufferHandle, clCreateImageFromFdINTEL, logical

xuguangxin commented on vpp/oclpostprocess_blender.cpp:
```diff
@@ -25,18 +25,182 @@
     if (status != YAMI_SUCCESS)
         return status;
+
+    blend(src, dest);
```
better have some return value

xuguangxin commented on vpp/oclpostprocess_blender.cpp:
```diff
@@ -25,18 +25,182 @@
+        ret = YAMI_FAIL;
+        goto err_cl;
+    }
+    bg_cl_mem_uv = clCreateImageFromFdINTEL(context, &import_info, &cl_status);
```
no need create it again, just bg_cl_mem_uv = dst_cl_mem_uv should ok.

xuguangxin commented on vpp/oclpostprocess_blender.cpp:
```diff
@@ -25,18 +25,182 @@
+    VASurfaceID src_id = (VASurfaceID)src->surface;
+    VASurfaceID dst_id = (VASurfaceID)dst->surface;
+    VAImage src_image, dst_image;
+    VABufferInfo src_info, dst_info;
```
We'd better use libyami's camel case name conversion?

dspmeng committed:
ocl: implement alpha blending

 * import cl image from VASurfaceID in ocl vpp base class
 * add opencl kernel for alpha blending

xuguangxin commented:
is it possible use  this one instead hard code path. 
https://www.khronos.org/registry/cl/sdk/1.0/docs/man/xhtml/clGetExtensionFunctionAddress.html

xuguangxin commented on vpp/oclpostprocess_blender.cpp:
```diff
@@ -25,18 +25,182 @@
+        ret = YAMI_FAIL;
+        goto err_cl;
+    }
+    clFlush(cmd_q);
```
do we really need this?

xuguangxin committed:
interface: add fourcc to VideoFrame

android's libva require this for csc

xuguangxin committed:
vpp: set color standard base on fourcc

android's libva require use set the color standard.

xuguangxin committed:
enc: base,h264,h265 init value before use it

xuguangxin committed:
basedec: bug fix, clear VideoFrame before use it

dspmeng commented:
Update according to review comments. Please review.

lizhong1008 committed:
h264enc: set sequence params for every frame

Set sequence params to driver for every frame,
else driver may crash when encoding some small size videos.

Signed-off-by: Zhong Li <zhong.li@intel.com>

lizhong1008 committed:
h264enc: set all reference lists to be short term ref

Because only short term reference supported right now,
set all reference pictures to be short term.

Signed-off-by: Zhong Li <zhong.li@intel.com>

xuguangxin commented on vpp/oclpostprocess_base.cpp:
```diff
@@ -56,6 +57,143 @@ YamiStatus OclPostProcessBase::ensureContext(const char* kernalName)
+    }
+    bufferInfo.mem_type = VA_SURFACE_ATTRIB_MEM_TYPE_DRM_PRIME;
+    if (VA_STATUS_SUCCESS != vaAcquireBufferHandle(m_display, clImage.image.buf, &bufferInfo)) {
+        ERROR("vaAcquireBufferHandle failed");
```
if we failed here. the image will leaked.

xuguangxin commented on vpp/oclpostprocess_base.cpp:
```diff
@@ -56,6 +57,143 @@ YamiStatus OclPostProcessBase::ensureContext(const char* kernalName)
 
+YamiStatus OclPostProcessBase::createCLImage(const SharedPtr<VideoFrame>& frame,
+                                             const cl_image_format& fmt,
+                                             OclImage& clImage)
```
could we use "const SharedPtr<OclImage>& clImage", so caller do not need care about release.

xuguangxin commented on vpp/oclpostprocess_blender.cpp:
```diff
@@ -25,19 +25,99 @@
+    localWorkSize[1] = 8;
+    globalWorkSize[0] = ALIGN16(dst.get()->crop.width) / pixelSize;
+    globalWorkSize[1] = ALIGN16(dst.get()->crop.height) / 2;
+    if (CL_SUCCESS != clEnqueueNDRangeKernel(m_context.get()->m_queue, m_kernel, 2, NULL,
```
m_context->m_queue is ok

xuguangxin commented on vpp/oclpostprocess_base.h:
```diff
@@ -34,6 +34,12 @@ class OclContext;
  *
  */
+typedef struct {
+    VAImage image;
```
better make this private. since only OclImage's deconstructor care about it

xuguangxin commented on vpp/oclpostprocess_base.cpp:
```diff
@@ -56,6 +57,143 @@ YamiStatus OclPostProcessBase::ensureContext(const char* kernalName)
+    return ret;
+}
+
+YamiStatus OclPostProcessBase::releaseCLImage(OclImage& clImage)
```
better define this in Ocl's deconstructor.

xuguangxin commented on vpp/oclpostprocess_base.cpp:
```diff
@@ -56,6 +57,143 @@ YamiStatus OclPostProcessBase::ensureContext(const char* kernalName)
+    cl_import_image_info_intel importInfo;
+    uint32_t height[3], i;
+
+    if (VA_STATUS_SUCCESS != vaDeriveImage(m_display, surfaceId, &clImage.image)) {
```
better use checkVaapiStatus

xuguangxin commented on vpp/oclpostprocess_base.cpp:
```diff
@@ -56,6 +57,143 @@ YamiStatus OclPostProcessBase::ensureContext(const char* kernalName)
+    for (int i = 0; i < clImage.planes; i++)
+        clReleaseMemObject(clImage.mem[i]);
+
+    vaReleaseBufferHandle(m_display, clImage.image.buf);
```
bette use checkVaapiStatus, many double release issue find by this.

lizhong1008 committed:
enc: make sure resolution is 16 aligned for vaSurface/vaContext

This patch is to fix android encoding driver crash when encoding
1920x1080 videos.

Signed-off-by: Zhong Li <zhong.li@intel.com>

lizhong1008 committed:
enc: add multi-reference-frame support.

1. Add multi-reference frames interface and test cases
2. Add multi-reference frames support for h264 and hevc.
   Because of current reoder mechanism, multi-reference is only used
   for forward reference. The max value of back referece frames is still
   one.

Signed-off-by: Zhong Li <zhong.li@intel.com>

dspmeng commented:
Address review comments and rebase to latest ocl_blend.

xuguangxin commented on vpp/oclpostprocess_base.cpp:
```diff
@@ -56,6 +58,124 @@ YamiStatus OclPostProcessBase::ensureContext(const char* kernalName)
     return YAMI_SUCCESS;
 }
 
+YamiStatus OclPostProcessBase::createCLImage(const SharedPtr<VideoFrame>& frame,
```
you can return SharedPtr<OclImage> directly.

xuguangxin commented on vpp/oclpostprocess_base.h:
```diff
@@ -45,7 +45,22 @@ class OclPostProcessBase : public IVideoPostProcess {
+
 protected:
+    struct OclImage {
+        VADisplay display;
```
better make this private

xuguangxin commented on vpp/oclpostprocess_base.h:
```diff
@@ -45,7 +45,22 @@ class OclPostProcessBase : public IVideoPostProcess {
 protected:
+    struct OclImage {
+        VADisplay display;
+        VAImage image;
```
better make this prviate

xuguangxin commented:
ok to me, please send pull request to 01.org/libyami
